fml/platform/darwin/scoped_nsobject.h

```diff
@@ -42,15 +42,37 @@ class scoped_nsprotocol {
  public:
   explicit scoped_nsprotocol(NST object = nil) : object_(object) {}
 
-  scoped_nsprotocol(const scoped_nsprotocol<NST>& that) : object_([that.object_ retain]) {}
+  scoped_nsprotocol(const scoped_nsprotocol<NST>& that)
```

Comment:
Handling of inclusion from arc and non-arc TUs is unrelated to AccessibilityBridge. Can you separate this out into its own patch and add a unit-test (in the fml_unittests harness) that ensures objects are correctly released. Crucially, I am not able to fully reason about the correctness of what happens when say an ARC TU vends an API to a non-ARC TU using this header and vice-versa.

cc @stuartmorgan who has experience with ARC migrations.

Reply:
Unfortunately it is not unrelated, it is necessary for this test to function.  Without it code that includes accessibility_bridge.h won't compile under ARC and OCMock breaks without ARC.

I could refactor the bridge to use the pimpl pattern but I think the right move is just to fix scoped_nsobject so we fix it everywhere once.

I can add a test that shows this works.

Reply:
> Unfortunately it is not unrelated, it is necessary for this test to function.

Can you clarify? I am not sure why the test cannot use manual reference counting in the test and compile the OCMock TUs with ARC. I'd like to understand why that causes an issue because I am afraid of introducing exactly the same kinds of issues here in the engine.

Reply:
Sure.  In my experience OCMock has problems in manually reference counted code.  I tried to find reference to the problems but couldn't pull anything up.  What it comes down to is that ARC has been adopted by the majority of developers and OCMock is doing funky things to work.  It just isn't as tested for manual reference counting.  I wouldn't consider it a concern for ARC adoption.  It is only an issue because of the odd nature of mocking libraries.

This change doesn't introduce ARC to the engine.  The engine will execute exactly the same.  It just allows us to continue writing our unit tests in ARC.  You'll have to take my word that OCMock works better with ARC.  This PR just makes our code ARC compatible, which is a positive move.

Reply:
Since there is still concern I added a unit-test.

Reply:
Have you looked at Chromium's current `scoped_ns*` implementations? Getting it right for cross-mode TUs was a nightmare, involving several iterations that turned out to have subtle undefined behavior bugs (once due to a non-obvious ODR violation, and once due to what eventually turned out to be a compiler bug IIRC).

Reply:
https://chromium.googlesource.com/chromium/src/base/+/master/mac/scoped_nsobject.h#49

Thanks Stuart, they use an extra template parameter which allows you to supply retain and release implementations (presumably they pass in noops in ARC code).

This PR is slightly different.  I first tried to to make scoped_nsobject work for ARC and non-ARC instances and ran into weirdness.  I suspect it was due to ARC's usage of the autorelease pool so that deallocation time isn't deterministic.

This doesn't have that problem because it just wraps objc message calls in c functions.  I didn't try to make scoped_nsobject usable with ARC instances.  In fact you get a linker error if you try to do that.  I'm only interested in letting ARC instances use objects that have scoped_nsobjects.  If I'm writing ARC C++ code I won't need scoped_nsobject, I can just rely on ARC.  The fact that this works is exercised in the new unit test I added.

As far as concerns about weirdness, this just wraps objc messages in c functions, otherwise behaves the same.  The PR passes integration tests and unit tests.  Since the only ARC code is unit tests, the worse case scenario is that our tests will flake or crash.

Reply:
> I'm only interested in letting ARC instances use objects that have scoped_nsobjects.

That's what we were doing too, and we hit UB trying.

Reply:
Do you have any documentation on what didn't work previously so we can see exactly what was tried and what went wrong?  Right now we have reason and empirical evidence that this works versus anecdotal evidence that something like it didn't work in a different context.

I appreciate the perspective.  The stakes are low here since this only affects testing targets.  Let's not hold back progress for a hypothetical, let's try to pin down the actual problem if one exists.